process/block/metablock.go

```diff
@@ -1561,39 +1561,120 @@ func (mp *metaProcessor) ApplyBodyToHeader(hdr data.HeaderHandler, bodyHandler d
-	endOfEpoch := &block.EndOfEpoch{
-		PendingMiniBlockHeaders: make([]block.ShardMiniBlockHeader, 0),
-		LastFinalizedHeaders:    make([]block.FinalizedHeaders, 0),
+func (mp *metaProcessor) createEpochStartForMetablock(metaBlock *block.MetaBlock) (*block.EpochStart, error) {
+	epochStart := &block.EpochStart{
+		LastFinalizedHeaders: make([]block.EpochStartShardData, 0),
 	}
 
-	if !mp.endOfEpochTrigger.IsEndOfEpoch() {
-		return endOfEpoch, nil
+	if !mp.epochStartTrigger.IsEpochStart() {
+		return epochStart, nil
 	}
 
-	pendingMiniBlocks := mp.pendingMiniBlocks.PendingMiniBlockHeaders()
-	endOfEpoch.PendingMiniBlockHeaders = pendingMiniBlocks
-
 	mp.mutNotarizedHdrs.RLock()
 	defer mp.mutNotarizedHdrs.RUnlock()
+
+	lastNotarizedHeaders := make([]data.HeaderHandler, mp.shardCoordinator.NumberOfShards())
 	for i := uint32(0); i < mp.shardCoordinator.NumberOfShards(); i++ {
 		lastNotarizedHdr := mp.lastNotarizedHdrForShard(i)
+		shardHdr, ok := lastNotarizedHdr.(*block.Header)
+		if !ok {
+			return nil, process.ErrWrongTypeAssertion
+		}
 
 		hdrHash, err := core.CalculateHash(mp.marshalizer, mp.hasher, lastNotarizedHdr)
 		if err != nil {
 			return nil, err
 		}
 
-		finalHeader := block.FinalizedHeaders{
-			ShardId:    lastNotarizedHdr.GetShardID(),
-			HeaderHash: hdrHash,
-			RootHash:   lastNotarizedHdr.GetRootHash(),
+		lastMetaHash, lastFinalizedMetaHash, err := mp.getLastFinalizedMetaHashForShard(shardHdr)
+		if err != nil {
+			return nil, err
+		}
+
+		finalHeader := block.EpochStartShardData{
+			ShardId:               lastNotarizedHdr.GetShardID(),
+			HeaderHash:            hdrHash,
+			RootHash:              lastNotarizedHdr.GetRootHash(),
+			FirstPendingMetaBlock: lastMetaHash,
+			LastFinishedMetaBlock: lastFinalizedMetaHash,
+		}
+
+		epochStart.LastFinalizedHeaders = append(epochStart.LastFinalizedHeaders, finalHeader)
+		lastNotarizedHeaders[i] = lastNotarizedHdr
+	}
+
+	pendingMiniBlocks, err := mp.pendingMiniBlocks.PendingMiniBlockHeaders(lastNotarizedHeaders)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, pendingMiniBlock := range pendingMiniBlocks {
+		recvShId := pendingMiniBlock.ReceiverShardID
+
+		currShardPendingMBs := epochStart.LastFinalizedHeaders[recvShId].PendingMiniBlockHeaders
+		currShardPendingMBs = append(currShardPendingMBs, pendingMiniBlock)
+	}
+
+	return epochStart, nil
+}
+
+func (mp *metaProcessor) getLastFinalizedMetaHashForShard(shardHdr *block.Header) ([]byte, []byte, error) {
+	var lastMetaHash []byte
+	var lastFinalizedMetaHash []byte
+
+	numAddedMetas := len(shardHdr.MetaBlockHashes)
+	if numAddedMetas > 0 {
+		lastMetaHash = shardHdr.MetaBlockHashes[numAddedMetas-1]
+	}
+
+	if numAddedMetas > 1 {
+		lastFinalizedMetaHash = shardHdr.MetaBlockHashes[numAddedMetas-2]
+		return lastMetaHash, lastFinalizedMetaHash, nil
+	}
+
+	for currentHdr := shardHdr; currentHdr.GetNonce() > 0 && currentHdr.GetEpoch() == shardHdr.GetEpoch(); {
```

Comment:
I think the lines 1625 to 1633 are not needed here and can be removed if you move lines [1636, 1639]  at line 1674 to set currentHdr instead of oldShardHdr and use currentHdr instead of oldShardHdr in this function. Then you would remove also lines [1641, 1644] and change the for condition to match.

Reply:
done.